audio/out/ao_netbsd.c

```diff
@@ -0,0 +1,280 @@
+#define NETBSD_MAX_CHANNELS (12)
+#endif
+
+#ifndef NETBSD_MIN_RATE
+#define NETBSD_MIN_RATE (1000)
+#endif
+
+#ifndef NETBSD_MAX_RATE
+#define NETBSD_MAX_RATE (192000)
+#endif
+
+struct priv {
+    int fd;
+    uint64_t total_blocks; /* audio blocks output */
+    uint64_t total_bytes; /* bytes sent to the queue */
+};
+
+static int init(struct ao *ao)
+{
+    char device[16] = "/dev/audio";
+    struct priv *p = ao->priv;
+    struct audio_info info;
+    struct audio_prinfo *pinfo;
+    struct mp_chmap_sel sel = {0};
+
+    AUDIO_INITINFO(&info);
+    pinfo = &info.play;
+
+    if (ao->device != NULL) {
+         (void)snprintf(device, sizeof(device), "/dev/audio%s",
+                        ao->device + sizeof("netbsd/") - 1);
+    }
+
+    MP_ERR(ao, "Opening device %s\n", device);
+    if ((p->fd = open(device, O_WRONLY)) == -1) {
+        MP_ERR(ao, "Can't open audio device %s: %s\n",
+               device, mp_strerror(errno));
+        goto fail;
+    }
+
+    info.mode = AUMODE_PLAY;
+
+    for (int n = 1; n <= NETBSD_MAX_CHANNELS; n++) {
+        struct mp_chmap map;
+
+        mp_chmap_from_channels(&map, n);
+        mp_chmap_sel_add_map(&sel, &map);
+    }
+
+    if (!ao_chmap_sel_adjust(ao, &sel, &ao->channels)) {
+        MP_ERR(ao, "Failed to select a valid channel map\n");
+        goto fail;
+    }
+
+    pinfo->channels = ao->channels.num;
+
+    if (ao->samplerate < NETBSD_MIN_RATE)
+        ao->samplerate = NETBSD_MIN_RATE;
+
+    if (ao->samplerate > NETBSD_MAX_RATE)
+        ao->samplerate = NETBSD_MAX_RATE;
+
+    pinfo->sample_rate = ao->samplerate;
+
+    switch (ao->format) {
+    case AF_FORMAT_U8:
+        pinfo->precision = 8;
+        pinfo->encoding = AUDIO_ENCODING_ULINEAR;
+        break;
+    case AF_FORMAT_S16:
+        pinfo->precision = 16;
+        pinfo->encoding = AUDIO_ENCODING_SLINEAR;
+        break;
+    default:
+        pinfo->precision = 32;
+        pinfo->encoding = AUDIO_ENCODING_SLINEAR;
+        ao->format = AF_FORMAT_S32;
+        break;
+    }
+
+    if (ioctl(p->fd, AUDIO_SETINFO, &info) == -1) {
+        MP_ERR(ao, "AUDIO_SETINFO failed: %s\n", mp_strerror(errno));
+        goto fail;
+    }
+
+    if (ioctl(p->fd, AUDIO_GETINFO, &info) == -1) {
+        MP_ERR(ao, "AUDIO_GETINFO failed: %s\n", mp_strerror(errno));
+        goto fail;
+    }
+
+    ao->samplerate = pinfo->sample_rate;
+
+    return 0;
+
+fail:
+    if (p->fd != -1) {
+        close(p->fd);
+    }
+    return -1;
+}
+
+static void uninit(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    (void)ioctl(p->fd, AUDIO_FLUSH, NULL);
+    (void)close(p->fd);
+    p->fd = -1;
+}
+
+static void reset(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+    struct audio_info info;
+    struct audio_offset offset;
+
+    if (ioctl(p->fd, AUDIO_GETINFO, &info) == -1) {
+        MP_ERR(ao, "AUDIO_GETINFO failed: %s\n", mp_strerror(errno));
+        return;
+    }
+
+    (void)ioctl(p->fd, AUDIO_DRAIN, NULL);
```

Comment:
Apparently that waits until the buffer plays. It's supposed to do the opposite. And we want (almost) never to block on the audio device either (in reset() we certainly do not).